Einsum

xdvriend opened the pull request:
**Short description**
This PR aims to complete and test the `einsum` API introduced by @lelemmen .

**Related issues**
closes #711 

**Additional context**
The first step was debugging the current implementation. By comparing it to the results from the same `einsum` operations using `numpy`, this should now be done.
What's next for this PR:

- [x]  Write appropriate tests for the einsum API.

- [x]  Add additional API, which follows `numpy` more closely. 

- [x] Check code for contractions and replace them with the `einsum` API where possible.

EDIT: Two of these marks have been done. However, I'd prefer it the way I mentioned in my comment below. To do this I will need the following:

- [x] ~~Possibility for dynamic tensor rank.~~
We discussed this and it would take up a lot of time without offering us much of an advantage in return.

**Don't forget:**
- [x] if new files are created: update the Doxygen file, the collective gqcp.hpp header and the CMake files

xdvriend commented:
@lelemmen I have added an initial implementation on th more `numpy`-like API. However, at the moment, it still requires a template argument `N`to be specified. Is there a way to circumvent this? If it is possible to calculate this `N`value within the second `einsum`method, I would be able to completely create the `tensor1.einsum("ijkl,jk->il", tensor2)` API.

xdvriend commented:
@lelemmen, I have noticed that the einsum API can't be used for something like `Tensor.einsum<2>("pq, qp ->", tensor2)`. Would you like me to provide this implementation as well, or would you leave this as it is now? 

Adding this implementation would allow me to rewrite several other methods using einsum, like `calculateElectronicEnergy()`of the Hartree-Fock methods and several `calculateExpectationValue()` methods. 

Other than that the einsum implementation should be finished.

I also added a method that transforms a tensor to a matrix, since this was often used in several methods, and I believed this made things a little clearer.

lelemmen commented:
I think your extra use case is quite useful, so could you add it to the code as well? Feel free to assign an extra estimate **directly on this PR**, indicating that this PR has become larger than the issue(s) it is trying to close.

xdvriend commented:
Subsequent implementations tackled in this PR:

- [ ]  Add ability to use einsum to contract over all axis indices.

- [ ] Replace `.contract()` with `.einsum` where possible, including contractions over all indices.

lelemmen commented:
> With these implementations done I believe the `contractWithMatrix()` API of the `SquareRankFourTensor` is now redundant, since `einsum()` has the possibility to contract a tensor with a matrix. Am I missing something or am I free to delete this method and replace it's uses with the corresponding `einsum()`(@lelemmen)?

Yes, you can delete this!

> EDIT: Upon further inspection, it seems the `contractWithMAtrix()` method is only used in `AtomicDecompositionParameters.cpp`, which looks to be disabled at the moment. Since it is very little work to replace these cases with an `einsum()` equivalent, I'm happy to do so. But, since testing is disabled for this module, I'm also happy to just leave it as it is now. Let me know what you prefer!

Indeed, this application is currently disabled. (See also #630, it can be re-enabled once the Mulliken partitioning works for two-electron operators). Anyways, I think you can leave that code as it is, since it's going to disappear anyways.

lelemmen commented on gqcp/include/Operator/SecondQuantized/SimpleSQTwoElectronOperator.hpp:
```diff
@@ -286,63 +283,30 @@ class SimpleSQTwoElectronOperator:
-                contraction_pair4);
-            result[i] = g_transformed;
+
+            // We will have to do four single contractions
+            // g(T U V W)  T^*(V R) -> a(T U R W)
+            // a(T U R W)  T(W S) -> b(T U R S)
+            const auto temp_1 = parameters[i].template einsum<1>("TUVW,VR->TURW", T_conjugate).template einsum<1>("TURW,WS->TURS", T_tensor);
+
+            // T(U Q)  b(T U R S) -> c(T Q R S)
+            const auto temp_2 = T_tensor.template einsum<1>("UQ,TURS->TQRS", temp_1);
+
+            // T^*(T P)  c(T Q R S) -> g'(P Q R S)
+            const auto transformed_component = T_conjugate.template einsum<1>("TP,TQRS->PQRS", temp_2);
+
+            result[i] = transformed_component;
         }
```
Looks a lot better, nice!!!

lelemmen commented on gqcp/include/QCModel/HF/GHF.hpp:
```diff
@@ -155,22 +153,20 @@ class GHF {
-        Eigen::TensorMap<Eigen::Tensor<const Scalar, 2>> P_aa_tensor {P_aa.data(), P_aa.rows(), P_aa.cols()};
-        Eigen::TensorMap<Eigen::Tensor<const Scalar, 2>> P_bb_tensor {P_bb.data(), P_bb.rows(), P_bb.cols()};
-
+        // Get the two-electron parameters
+        const auto& g = sq_hamiltonian.twoElectron().parameters();
```
```suggestion
```

I've moved this again to below.

lelemmen commented on gqcp/include/QCModel/HF/GHF.hpp:
```diff
@@ -200,30 +196,21 @@ class GHF {
-        Eigen::TensorMap<Eigen::Tensor<const Scalar, 2>> P_ba_tensor {P_ba.data(), P_ba.rows(), P_ba.cols()};
-        Eigen::TensorMap<Eigen::Tensor<const Scalar, 2>> P_bb_tensor {P_bb.data(), P_bb.rows(), P_bb.cols()};
-
+        // Get the two-electron parameters
+        const auto& g = sq_hamiltonian.twoElectron().parameters();
```
```suggestion
```

I've moved this to below.

lelemmen commented on gqcp/include/QCModel/HF/GHF.hpp:
```diff
@@ -200,30 +196,21 @@ class GHF {
-        Eigen::Map<Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic>> K_ba {K_ba_tensor.data(), K_ba_tensor.dimension(0), K_ba_tensor.dimension(1)};
-        Eigen::Map<Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic>> K_bb {K_bb_tensor.data(), K_bb_tensor.dimension(0), K_bb_tensor.dimension(1)};
-
+        Tensor<Scalar, 2> K_aa_tensor = g.template einsum<2>("ijkl,kj->il", P_aa);
+        Tensor<Scalar, 2> K_ab_tensor = g.template einsum<2>("ijkl,kj->il", P_ba);
+        Tensor<Scalar, 2> K_ba_tensor = g.template einsum<2>("ijkl,kj->il", P_ab);
+        Tensor<Scalar, 2> K_bb_tensor = g.template einsum<2>("ijkl,kj->il", P_bb);
+
+        // Convert the given tensors back to a matrix.
+        auto K_aa = K_aa_tensor.toMatrix(K_aa_tensor.dimension(0), K_aa_tensor.dimension(1));
+        auto K_ab = K_ab_tensor.toMatrix(K_ab_tensor.dimension(0), K_ab_tensor.dimension(1));
+        auto K_ba = K_ba_tensor.toMatrix(K_ba_tensor.dimension(0), K_ba_tensor.dimension(1));
+        auto K_bb = K_bb_tensor.toMatrix(K_bb_tensor.dimension(0), K_bb_tensor.dimension(1));
```
```suggestion
        const auto& g = sq_hamiltonian.twoElectron().parameters();
        const auto K_aa= g.template einsum<2>("ijkl,kj->il", P_aa).asMatrix();
        const auto K_ab = g.template einsum<2>("ijkl,kj->il", P_ba).asMatrix();
        const auto K_ba = g.template einsum<2>("ijkl,kj->il", P_ab).asMatrix();
        const auto K_bb = g.template einsum<2>("ijkl,kj->il", P_bb).asMatrix();
```

lelemmen commented on gqcp/include/Mathematical/Representation/Tensor.hpp:
```diff
@@ -287,37 +291,124 @@ class Tensor:
+    }
+
+
+    /**
+     *  Convert the given tensor to a matrix with the specified dimensions
+     *
+     *  @param rows        the dimension of the rows of the output matrix
+     *  @param cols        the dimension of the columns of the output matrix
+     */
+    const GQCP::Matrix<Scalar> toMatrix(const size_t rows, const size_t cols) const {
+
+        const auto rank_two_tensor = *this;
+        return GQCP::Matrix<Scalar>(Eigen::Map<const Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic>>(rank_two_tensor.data(), rows, cols));
     }
```
I think you can clean up this API even more. Could you also add the appropriate `enable_if`?

```suggestion
    /**
     *  @return This rank-two tensor as a matrix.
     */
    const GQCP::Matrix<Scalar> asMatrix() const {

        const auto rows = this->dimension(0);
        const auto cols = this->dimension(1);

        return GQCP::Matrix<Scalar>(Eigen::Map<const Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic>>(this->data(), rows, cols));
    }
```

lelemmen commented on gqcp/include/QCModel/HF/GHF.hpp:
```diff
@@ -155,22 +153,20 @@ class GHF {
-        // Do the actual contractions, and convert the given tensor back to a matrix.
-        const auto& g = sq_hamiltonian.twoElectron().parameters();
-        Tensor<Scalar, 2> J_tensor = P_aa_tensor.contract(g.Eigen(), direct_contraction_pair) + P_bb_tensor.contract(g.Eigen(), direct_contraction_pair);
+        Tensor<Scalar, 2> Ja_tensor = g.template einsum<2>("ijkl,kl->ij", P_aa);
+        Tensor<Scalar, 2> Jb_tensor = g.template einsum<2>("ijkl,kl->ij", P_bb);
```
```suggestion
        const auto Ja_tensor = g.template einsum<2>("ijkl,kl->ij", P_aa).asMatrix();
        const auto Jb_tensor = g.template einsum<2>("ijkl,kl->ij", P_bb).asMatrix();
```

lelemmen commented on gqcp/include/QCModel/HF/GHF.hpp:
```diff
@@ -155,22 +153,20 @@ class GHF {
+        Tensor<Scalar, 2> Jb_tensor = g.template einsum<2>("ijkl,kl->ij", P_bb);
 
-        Eigen::Map<Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic>> J {J_tensor.data(), J_tensor.dimension(0), J_tensor.dimension(1)};
+        // Convert the given tensors back to a matrix.
+        auto Ja = Ja_tensor.toMatrix(Ja_tensor.dimension(0), Ja_tensor.dimension(1));
+        auto Jb = Jb_tensor.toMatrix(Jb_tensor.dimension(0), Jb_tensor.dimension(1));
```
```suggestion
```

lelemmen commented on gqcp/tests/Mathematical/Representation/Tensor_test.cpp:
```diff
@@ -281,3 +281,105 @@ BOOST_AUTO_TEST_CASE(addBlock_matrix) {
         }
     }
 }
+
```
Add documentation before the test, like you would for a function.

lelemmen commented on gqcp/tests/Mathematical/Representation/Tensor_test.cpp:
```diff
@@ -281,3 +281,105 @@ BOOST_AUTO_TEST_CASE(addBlock_matrix) {
+            }
+        }
+    }
+
+    GQCP::MatrixX<double> reference_block_1 {2, 2};
+    // clang-format off
+    reference_block_1 << 50, 60,
+                         60, 74;
+    // clang-format on
+
+    GQCP::MatrixX<double> reference_block_2 {2, 2};
+    // clang-format off
+    reference_block_2 <<   55,   67,
+                           65,   81;
+    // clang-format on
+
+    reference_tensor.addBlock<2, 3>(reference_block_1, 0, 0, 0, 0);
+    reference_tensor.addBlock<2, 3>(reference_block_1, 0, 1, 0, 0);
+    reference_tensor.addBlock<2, 3>(reference_block_2, 1, 0, 0, 0);
+    reference_tensor.addBlock<2, 3>(reference_block_2, 1, 1, 0, 0);
+
+    const auto output4 = T1.einsum<1>(T2, "ijkl", "ia", "jkla");
+    const auto output5 = T1.einsum<1>("ijkl,ia->jkla", T2);
+    const auto output6 = T1.einsum<1>("ijkl,ia->jkla", M);
+
+    BOOST_CHECK(reference_tensor.isApprox(output4, 1e-12));
+    BOOST_CHECK(reference_tensor.isApprox(output5, 1e-12));
+    BOOST_CHECK(reference_tensor.isApprox(output6, 1e-12));
+
+    // Let's test a contraction over all indices as well
+    // We test both einsum notations and the contraction of a tensor with a matrix
+    // reference values were calculated with numpy.einsum()
+    const auto output7 = T1.einsum<4>(T1, "ijkl", "ijkl", "");
+    const auto output8 = T1.einsum<4>("ijkl,ijkl->", T1);
+    const auto output9 = T2.einsum<2>("ij,ij->", M);
+
+    BOOST_CHECK_EQUAL(output7(0), 1096);
+    BOOST_CHECK_EQUAL(output8(0), 1096);
+    BOOST_CHECK_EQUAL(output9(0), 62);
+}
```
Check the punctuation and capitalization of the documentation, please.

Also, I see that you're testing multiple things: it's better split this up in different test cases, each specifically testing one part of the API.

lelemmen reviewed (changes requested):
This is really awesome!

For the review, I've added some suggestions on improving the `asMatrix()` API. Also, The RHF and UHF QCModel files should be cleaned up like I did in the GHF file.

There's still some clean-up to be done. I'd like to ask you to be more diligent about your documentation (they're supposed to be real sentences).

xdvriend commented on gqcp/include/Mathematical/Representation/Tensor.hpp:
```diff
@@ -287,37 +291,124 @@ class Tensor:
-                intermediate_indices.erase(match);
-                intermediate_indices.erase(i);
+                intermediate_indices.erase(match, 1);
+                intermediate_indices.erase(i - 1, 1);
```
Exactly! And I'll be sure to add it.

xdvriend commented on gqcp/include/Mathematical/Representation/Tensor.hpp:
```diff
@@ -287,37 +291,124 @@ class Tensor:
+     *  @return The result of the tensor contraction.
+     */
+    template <int N, int LHSRank = Rank, int RHSRank>
+    Tensor<Scalar, LHSRank + RHSRank - 2 * N> einsum(std::string contraction_string, const Tensor<Scalar, RHSRank>& rhs) const {
```
The `contraction_string`cannot be `const`since it is adjusted by the boost library when removing spaces and symbols.

xdvriend commented:
I have adjusted everything as you mentioned it and replaced the final instances of the old `contractWithMatrix()` API. The only thing I didn't change was the following piece of code from the RHF `QCModel`:

![Screenshot 2020-11-10 at 17 28 07](https://user-images.githubusercontent.com/54795591/98702086-483ef580-237a-11eb-9f6d-133a0afcb775.png)

- I'm not using `const auto` since it ill automatically recognise it as an `Eigen::Tensor`, which can't use the `.asMatrix()` method.
- I know I should be able to write it in one line, by adding `.asMatrix()` at the end of the einsum, but for some reason, this causes the `properties_test.cpp`and `constrained_RHF_test.cpp` to fail. Because of this I have left it as it is now.

lelemmen reviewed (approved):
Your remarks sounds good to me. This PR is good to go! 🎉
